### PR TITLE
BUG: Parquet does not support saving fp16 GH#44846

### LINKED ISSUE #44846: BUG:  Parquet format does not support saving float16 columns
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np

data = np.arange(2, 10, dtype=np.float16)
df = pd.DataFrame(data=data, columns=['fp16'])
df.to_parquet('./fp16.parquet')
```

### Issue Description

Pandas does not validate presence of float16 columns in DataFrame as parquet format does not support saving float 16 values.

Additional references:
- https://github.com/apache/arrow/issues/2691
- https://stackoverflow.com/questions/54150865/arrownotimplementederror-halffloat-error-on-applying-pandas-to-feather-on-a-dat

### Expected Behavior

Raise ValueError while validating DataFrame before saving.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : b7991da3612de3f8119841d46f95c23148b0d928
python           : 3.8.12.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.11.0-41-generic
Version          : #45~20.04.1-Ubuntu SMP Wed Nov 10 10:20:10 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.4.0.dev0+1356.gb7991da361.dirty
numpy            : 1.21.4
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.3.1
setuptools       : 59.4.0
Cython           : 0.29.25
pytest           : 6.2.5
hypothesis       : 6.31.3
sphinx           : 4.3.1
blosc            : None
feather          : None
xlsxwriter       : 3.0.2
lxml.etree       : 4.6.4
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 3.0.3
IPython          : 7.30.1
pandas_datareader: None
bs4              : 4.10.0
bottleneck       : 1.3.2
fsspec           : 2021.11.0
fastparquet      : 0.7.2
gcsfs            : 2021.11.0
matplotlib       : 3.5.0
numexpr          : 2.8.0
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 6.0.1
pyxlsb           : None
s3fs             : 2021.11.0
scipy            : 1.7.3
sqlalchemy       : 1.4.28
tables           : 3.6.1
tabulate         : 0.8.9
xarray           : 0.18.2
xlrd             : 2.0.1
xlwt             : 1.3.0
numba            : 0.53.1

</details>

### PR DESCRIPTION
**Pull request by Anirudhsekar96**
- [x] closes #44846
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] Adds validation check for float16 column in to_parquet function by raising a ValueError

### CODE REVIEW

File: pandas/io/parquet.py [hunk 1]
```diff
@@ -136,6 +136,10 @@ def validate_dataframe(df: DataFrame):
         if not valid_names:
             raise ValueError("Index level names must be strings")
 
+        # Paquet format does not support saving float16
```
Comment (jreback) [inline pandas/io/parquet.py:139]: well does this actually fail in pyarrow or fastparquet?

we barely support float16 in pandas anyhow

**Review by jreback — CHANGES_REQUESTED** [review]
would need a test and release note

File: pandas/io/parquet.py (hunk 1 above)
Comment (Anirudhsekar96) [reply pandas/io/parquet.py]: It fails in PyArrow, Parquet format does not support saving fp16 files but fastparquet seems to typecast fp16 to fp32 before saving.

Release Note: 'Added validation checks for float16 dtype in pandas.DataFrame.to_parquet(engine="pyarrow")'

**Comment by jorisvandenbossche** [comment]
Is this explicit check needed? It seems pyarrow itself already gives a decent error message?

```
In [4]: import pandas as pd
   ...: import numpy as np
   ...: 
   ...: data = np.arange(2, 10, dtype=np.float16)
   ...: df = pd.DataFrame(data=data, columns=['fp16'])
   ...: df.to_parquet('./fp16.parquet')
...
ArrowNotImplementedError: Unhandled type for Arrow to Parquet schema conversion: halffloat
```

(adding an explicit check also means we wouldn't automatically get the improvement if this is fixed on pyarrow's side)

**Comment by Anirudhsekar96** [comment]
In the given example, PyArrow creates a file called 'fp16.parquet' before throwing the error resulting in an empty file. An explicit check before passing to PyArrow would prevent the creation of the empty file.

Additionally, it seems that there is little interest in adding support for float16 in PyArrow at the moment (see ref: https://github.com/apache/arrow/issues/2691). Parquet format also does not seem to have adding extensions for supporting float16 in the road map (ref: https://issues.apache.org/jira/browse/ARROW-7242). 

Another option would be to coerce all the columns with float16 coerce to float32 and adding a warning indicating the conversion. Currently fastparquet handles fp16 using the conversion method.

File: pandas/io/parquet.py [hunk 2]
```diff
@@ -167,6 +167,10 @@ def write(
     ):
         self.validate_dataframe(df)
 
+        # PyArrow does not support saving float16
```
Comment (jreback) [inline pandas/io/parquet.py:170]: comment is superfluous

Comment (jreback) [reply pandas/io/parquet.py:170]: can you add a more meaningful message e.g. `columns [.....]` have this dtype.

File: pandas/tests/io/test_parquet.py [hunk 3]
```diff
@@ -707,6 +707,14 @@ def test_duplicate_columns(self, pa):
         df = pd.DataFrame(np.arange(12).reshape(4, 3), columns=list("aaa")).copy()
         self.check_error_on_write(df, pa, ValueError, "Duplicate column names found")
 
+    def test_write_column_fp16(self, pa):
```
Comment (jreback) [inline pandas/tests/io/test_parquet.py:710]: def test_unsupported_float16

**Review by jreback — CHANGES_REQUESTED** [review]
i am ok with this, a partial write is not a great end result here.

could you add a note in I/O section for 1.4.0

File: pandas/io/parquet.py (hunk 2 above)
Comment (Anirudhsekar96) [reply pandas/io/parquet.py]: Thanks, removed the comment and changed the message to include column names with dtype as float 16.

File: pandas/tests/io/test_parquet.py (hunk 3 above)
Comment (Anirudhsekar96) [reply pandas/tests/io/test_parquet.py]: Thanks, I have changed the name of the test in the latest commit.

**Comment by jorisvandenbossche** [comment]
> In the given example, PyArrow creates a file called 'fp16.parquet' before throwing the error resulting in an empty file. An explicit check before passing to PyArrow would prevent the creation of the empty file.

That might actually be a bug in the pandas code, as I can't reproduce it with just pyarrow (I do reproduce it with the pandas version above):

```python
import pyarrow as pa
import pyarrow.parquet as pq

table = pa.table({'a': np.array([0, 1, 2], dtype="float16")})
pq.write_table(table, "test_halffloat.parquet")
```

**Comment by jorisvandenbossche** [comment]
So this "creates emtpy file" issue also happens for other data types that aren't supported with Parquet (such as timedelta64, or for any other reason that pyarrow raises an error while writing). I think that is something we should fix in general instead of doing a special case check for float16.

**Comment by jreback** [comment]
good point @jorisvandenbossche 

ok let's close this issue and open one for the writing partial files

**Comment by Anirudhsekar96** [comment]
Opened a new bug report for the broader issue:

https://github.com/pandas-dev/pandas/issues/44914

**Pull request closed without merging**